### DISCUSSION TITLE
How to know which dependency file (.whl/.tar.gz) Poetry used for the dependency installation?

### DISCUSSION THREAD

**Discussion by talarian1** [python-poetry/poetry#6869] [Q&A]
I'm one of the maintainers of JFrog CLI and a few other JFrog tools used by developers.
We have recently added support for creating [Build Info](https://www.buildinfo.org/) for Poetry projects. The build info includes the list of project modules, artifacts, dependencies, environment variables, and more. One of the key features build info offers to the community is the knowledge of which dependencies were installed as part of the build, since build info is practically a precise recording of the build. Our current support for Poetry is no longer working due to this breaking change #6356.
Any suggestion on how we can find out the dependencies’ sha2 or the .whl/.tar.gz file system location that was used during the Poetry installation?
Our users' community would really like to see Poetry included in JFrog CLI's build info offering, and it would be a shame to not have it there. Note that other package managers like pip and pipenv are already supported.
Since the Poetry code decides which dependency to download, will you be able to show us where this code is, so we can use it to retrieve this information?
Thanks in advance :)

**Comment by radoering** [comment]
This should be the place where poetry decides which distribution to choose for a specific package: https://github.com/python-poetry/poetry/blob/16046d9ac9b72a49e1bc4618fb686695cc64821c/src/poetry/installation/chooser.py#L72-L108

Be aware that poetry's internal code may change from one release to the next. A more sustainable way might be to add an option to the `install` command to output the filename/sha of the chosen distributions.

Btw, writing the direct_url.json for index dependencies was a bug (see [Recording the Direct URL Origin of installed distributions](https://packaging.python.org/en/latest/specifications/direct-url/#recording-the-direct-url-origin-of-installed-distributions)):

> This file MUST NOT be created when installing a distribution from an other type of requirement (i.e. name plus version specifier).

Can you provide information how you extract the required information from pip or pipenv. At least, pip doesn't create a direct_url.json for index dependencies, does it?

**Reply by talarian1** [reply]
Hi, thanks for the quick response.
A flag to the poetry `install` command to output the filename/sha of the chosen distributions - sounds exactly what we need.
Both Pip and pipenv print this information when the verbose flag is supplied to the install command.

**Reply by neersighted** [reply]
I'd personally suggest some sort of `install --json` flag that outputs a report of installed/removed/changed dists after all operations complete.

**Reply by rvodden** [reply]
I've [started to pull together a poetry plugin which w](https://github.com/HumanisingAutonomy/poetry_buildinfo)ould provide this information. @neersighted I can't work out how to add an option to an existing command with a plugin - i tried extending the InstallCommand object, however when I register it, cleo complains that the "install" command is already registered. For the moment I've extended the executor and am injecting that into the command at install and update events. This is at least working, so I can work on the core functionality. All feedback, especially feedback about how I can better hook into Poetry, is very welcome :-)

**Comment by eyalbe4** [comment]
@rvodden - Thanks so much for working on this plugin!
@neersighted - Any assistance that you cam provide @rvodden weould be greatly appreciated!
As @talarian1 mentioned above, we'd really like to see JFrog CLI supporting build-info for Poetry.

**Reply by neersighted** [reply]
For assistance on plugin authorship, I would suggest reaching out on Discord. The faster round trips help a lot when it comes to developing code.

**Reply by rvodden** [reply]
I've just pushed up a new version which demonstrates that what we're trying to achieve is feasible. This creates a file called build-log.json in the current working directory with what I believe to a be a suitable buildinfo json object. Still to do are
* documentation (sigh)
* more tests - I've inherited from the executor object in poetry, so want to port over any tests which hit the code i've changed
* the file needs to be written to the jfrog temporary directory which is used to collect buildinformation prior to publication - its not immediately obvious how to work out where that directory is
* i'm reasonably sure that if any dependencies are not pulled from pypi via artifactory then right now the plugin will fail - this will be detected when I port over the tests though, and I'll resolve after that

It would be great if @eyalbe4 you could do a quick test? The easiest way to install it is to have a virtual environment which is used by a project of yours, and this project. Run poetry install in this project and then poetry install in your project, and you should get a build-log.json file. `--no-plugins` is an extremely helpful option on poetry if stuff gets tied in knots!

**Comment by eyalbe4** [comment]
Hi @rvodden,
Sorry for the late response and thank you for your interest in this.
Our initial goal was to introduce [build-info](https://github.com/jfrog/build-info-go) support for Poetry, but at the end of the development effort, we were forced to stop due to the issue discussed here. As much we want to get back to this (we feel this addition is important), our current focus was shifted to other areas. Hopefully we'll be able to get back to this in the near future.
